src/tree.rs

```diff
@@ -3,79 +3,270 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 use std::{cmp::{Eq, PartialEq},
-          collections::HashSet,
+          collections::{HashMap, HashSet},
           fmt,
           hash::{Hash, Hasher},
-          iter};
+          ops::Deref};
 
-/// Synced item kinds. Each corresponds to a Sync record type.
-#[derive(Eq, Hash, PartialEq)]
-pub enum Kind {
-    Bookmark,
-    Query,
-    Folder,
-    Livemark,
-    Separator,
+/// A complete, rooted bookmark tree with tombstones.
+///
+/// The tree stores bookmark nodes in a vector, and uses indices in the vector
+/// to identify parents and children. This makes traversal and lookup very
+/// efficient. Retrieving a node's parent takes one indexing operation,
+/// retrieving children takes one indexing operation per child, and retrieving a
+/// node by random GUID takes one hash map lookup and one indexing operation.
+#[derive(Debug)]
+pub struct Tree {
+    entries: Vec<Entry>,
+    index_by_guid: HashMap<String, usize>,
+    deleted_guids: HashSet<String>,
 }
 
-impl fmt::Display for Kind {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        f.write_str(match self {
-                        Kind::Bookmark => "Bookmark",
-                        Kind::Query => "Query",
-                        Kind::Folder => "Folder",
-                        Kind::Livemark => "Livemark",
-                        Kind::Separator => "Separator",
-                    })
+impl Tree {
+    /// Constructs a new rooted tree.
+    pub fn new(root: Item) -> Tree {
+        let mut index_by_guid = HashMap::new();
+        index_by_guid.insert(root.guid.to_string(), 0);
+        Tree { entries: vec![Entry { parent_index: None,
+                                     item: root,
+                                     level: 0,
+                                     child_indices: Vec::new(), }],
+               index_by_guid,
+               deleted_guids: HashSet::new(), }
+    }
+
+    pub fn deletions(&self) -> Vec<&str> {
```

Comment:
I think you should be able to return the iterator like this? Would be nice to avoid the unnecessary collect.

```rust
pub fn deletions<'a>(&'a self) -> Iterator<Item = &str> + 'a {
    self.deleted_guids
        .iter()
        .map(move |guid| guid.as_ref())
}
```